src/world_diff.rs

```diff
@@ -227,27 +268,23 @@ impl WorldDiff {
             l2_to_l1_logs: self.l2_to_l1_logs.snapshot(),
             transient_storage_changes: self.transient_storage_changes.snapshot(),
             pubdata: self.pubdata.snapshot(),
+            storage_refunds: self.storage_refunds.snapshot(),
+            pubdata_costs: self.pubdata_costs.snapshot(),
         }
     }
 
-    pub(crate) fn rollback(
-        &mut self,
-        Snapshot {
-            storage_changes,
-            paid_changes,
-            events,
-            l2_to_l1_logs,
-            transient_storage_changes,
-            pubdata,
-        }: Snapshot,
-    ) {
-        self.storage_changes.rollback(storage_changes);
-        self.paid_changes.rollback(paid_changes);
-        self.events.rollback(events);
-        self.l2_to_l1_logs.rollback(l2_to_l1_logs);
+    pub(crate) fn rollback(&mut self, snapshot: Snapshot, rollback_io_history: bool) {
+        self.storage_changes.rollback(snapshot.storage_changes);
+        self.paid_changes.rollback(snapshot.paid_changes);
+        self.events.rollback(snapshot.events);
+        self.l2_to_l1_logs.rollback(snapshot.l2_to_l1_logs);
         self.transient_storage_changes
-            .rollback(transient_storage_changes);
-        self.pubdata.rollback(pubdata);
+            .rollback(snapshot.transient_storage_changes);
+        self.pubdata.rollback(snapshot.pubdata);
+        if rollback_io_history {
```

Comment:
Looks like this only happens in the external rollback, so please remove the boolean and move the if's body to the external rollback.